### PR TITLE
FIPS: add indicator for PKCS#1 version 1.5 padding usage for key transport and agreement

### PR DESCRIPTION
**Pull request by paulidale**
As mandated in [NIST SP 800-131A Rev. 2](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-131Ar2.pdf) section 6, PKCS#1 version 1.5 padding is no longer permitted for key agreement and key transport using RSA.

This is implemented here as a hard failure condition.  There is also a FIPS provider parameter that permits the use of this padding with an indicator.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: apps/fipsinstall.c [hunk 1]
```diff
@@ -92,8 +92,10 @@ const OPTIONS fipsinstall_options[] = {
      "Disallow DSA signing"},
     {"tdes_encrypt_disabled", OPT_DISALLOW_TDES_ENCRYPT, '-',
      "Disallow Triple-DES encryption"},
+    {"rsa_pkcs15_padding_disabled", OPT_DISALLOW_PKCS15_PADDING, '-',
+     "Disallow PKCS#1 version 1.5 padding for RSA encryption"},
     {"rsa_sign_x931_disabled", OPT_DISALLOW_SIGNATURE_X931_PADDING, '-',
-     "Disallow X931 Padding for RSA signing"},
+     "Disallow X9.31 padding for RSA signing"},
```
Comment (slontis) [inline apps/fipsinstall.c:98]: signing?

Comment (paulidale) [reply apps/fipsinstall.c:98]: Oops.

File: include/openssl/fips_names.h [hunk 2]
```diff
@@ -141,6 +141,10 @@ extern "C" {
  */
 # define OSSL_PROV_FIPS_PARAM_TDES_ENCRYPT_DISABLED "tdes-encrypt-disabled"
 
+/*
```
Comment (slontis) [inline include/openssl/fips_names.h:144]: Comment?

File: providers/implementations/asymciphers/rsa_enc.c [hunk 3]
```diff
@@ -462,6 +475,9 @@ static int rsa_set_ctx_params(void *vprsactx, const OSSL_PARAM params[])
     if (!OSSL_FIPS_IND_SET_CTX_PARAM(prsactx, OSSL_FIPS_IND_SETTABLE0, params,
                                      OSSL_ASYM_CIPHER_PARAM_FIPS_KEY_CHECK))
         return  0;
+    if (!OSSL_FIPS_IND_SET_CTX_PARAM(prsactx, OSSL_FIPS_IND_SETTABLE1, params,
+                                     OSSL_ASYM_CIPHER_PARAM_PKCS15_PADDING_DISABLED))
```
Comment (slontis) [inline providers/implementations/asymciphers/rsa_enc.c:479]: OSSL_ASYM_CIPHER_PARAM_PKCS15_PADDING_DISABLE Needs to be in the settables list

File: include/openssl/fips_names.h [hunk 4]
```diff
@@ -141,6 +141,14 @@ extern "C" {
...
+ * A boolean that determines if RSA key agreement and transport operations are
```
Comment (slontis) [inline include/openssl/fips_names.h:145]: Not quite right :)

File: include/openssl/fips_names.h [hunk 5]
```diff
@@ -142,8 +142,9 @@ extern "C" {
 # define OSSL_PROV_FIPS_PARAM_TDES_ENCRYPT_DISABLED "tdes-encrypt-disabled"
 
 /*
- * A boolean that determines if RSA key agreement and transport operations are
- * allowed.  See SP800-131A r2 for further information.
+ * A boolean that determines if PKCS#1 v1.5 padding is allowed for key
```
Comment (slontis) [inline include/openssl/fips_names.h:145]: please remove the last allowed bit

**Review by slontis — APPROVED** [review]

**Comment by slontis** [comment]
@t8m are you ok with the way the tests are done here for tls?
Disabling RSA encryption (which uses PKCS1.5), pretty much breaks a lot of TLS tests.

**Comment by paulidale** [comment]
Clean rebase to master

**Comment by paulidale** [comment]
Ping @openssl/committers for second review please.

File: apps/cms.c
Comment (beldmit) [inline apps/cms.c:1]: The changes in this file seem unrelated to the PR

Comment (paulidale) [reply apps/cms.c:1]: They are formatting fixes I did while trying to figure out why the CMS tests failed.
Happy to drop this if you want.

**Review by beldmit — APPROVED** [review]
LGTM. I'd drop the `apps/cms.c` from the PR and provide them separately but it's nt crucial

File: apps/cms.c
Comment (paulidale) [reply apps/cms.c:1]: I've dropped this.  No need for reapprovals IMO.

**Comment by t8m** [comment]
@beldmit @slontis please reconfirm as this was rebased.

**Comment by paulidale** [comment]
We'll have to rebase most of these when the merge.  They all change the exact same line in fipsprov.c 😞

**Comment by mattcaswell** [comment]
If its a trivial rebase I don't think a reconfirm is necessary.

**Comment by t8m** [comment]
AFAIK there were more changes since the approval from @slontis. However I assume @beldmit approval still holds then.

**Comment by paulidale** [comment]
This one also dropped the CMS formatting changes: three new lines and a similar number of != NULL.

**Comment by paulidale** [comment]
@beldmit @slontis ?

File: providers/implementations/asymciphers/rsa_enc.c [hunk 6]
```diff
@@ -158,6 +158,19 @@ static int rsa_encrypt(void *vprsactx, unsigned char *out, size_t *outlen,
     if (!ossl_prov_is_running())
         return 0;
 
+#ifdef FIPS_MODULE
+    if ((prsactx->pad_mode == RSA_PKCS1_PADDING
+         || prsactx->pad_mode == RSA_PKCS1_WITH_TLS_PADDING)
+        && !ossl_FIPS_IND_on_unapproved(OSSL_FIPS_IND_GET(prsactx),
```
Comment (slontis) [inline providers/implementations/asymciphers/rsa_enc.c:164]: any reason to not use the macro here?

**Review by slontis — COMMENTED** [review]
Still approved with or without my comment

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

File: providers/implementations/asymciphers/rsa_enc.c (hunk 6 above)
Comment (paulidale) [reply providers/implementations/asymciphers/rsa_enc.c:164]: Copied from somewhere else I think.  Worth fixing in due course.

Comment (slontis) [reply providers/implementations/asymciphers/rsa_enc.c:164]: FYI - It only does this in a few places in order to share the code

**Pull request closed without merging**